From 6c85edcecb539dd6e294aeea83578b1469b5e907 Mon Sep 17 00:00:00 2001
From: Nikolay Aleksandrov <nikolay@redhat.com>
Date: Fri, 5 Sep 2014 11:03:55 -0400
Subject: [ethernet] sfc: Use pci_enable_msix_range() instead of pci_enable_msix()

Message-id: <1409915048-26438-20-git-send-email-nikolay@redhat.com>
Patchwork-id: 91413
O-Subject: [RHEL7.1 PATCH BZ 1110888 19/32] sfc: Use pci_enable_msix_range() instead of pci_enable_msix()
Bugzilla: 1110888
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Alexander Gordeev <agordeev@redhat.com>

As result of deprecation of MSI-X/MSI enablement functions
pci_enable_msix() and pci_enable_msi_block() all drivers
using these two interfaces need to be updated to use the
new pci_enable_msi_range() and pci_enable_msix_range()
interfaces.

Signed-off-by: Alexander Gordeev <agordeev@redhat.com>
Cc: Shradha Shah <sshah@solarflare.com>
Cc: linux-net-drivers@solarflare.com
Cc: netdev@vger.kernel.org
Cc: linux-pci@vger.kernel.org
Acked-by: Shradha Shah <sshah@solarflare.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 184603d8825773afae24279e055440dddada478d)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/sfc/efx.c b/drivers/net/ethernet/sfc/efx.c
index a042898..aacc56d 100644
--- a/drivers/net/ethernet/sfc/efx.c
+++ b/drivers/net/ethernet/sfc/efx.c
@@ -1344,20 +1344,23 @@ static int efx_probe_interrupts(struct efx_nic *efx)
 
   for (i = 0; i < n_channels; i++)
    xentries[i].entry = i;
-  rc = pci_enable_msix(efx->pci_dev, xentries, n_channels);
-  if (rc > 0) {
+  rc = pci_enable_msix_range(efx->pci_dev,
+        xentries, 1, n_channels);
+  if (rc < 0) {
+   /* Fall back to single channel MSI */
+   efx->interrupt_mode = EFX_INT_MODE_MSI;
+   netif_err(efx, drv, efx->net_dev,
+      "could not enable MSI-X\n");
+  } else if (rc < n_channels) {
    netif_err(efx, drv, efx->net_dev,
       "WARNING: Insufficient MSI-X vectors"
       " available (%d < %u).\n", rc, n_channels);
    netif_err(efx, drv, efx->net_dev,
       "WARNING: Performance may be reduced.\n");
-   EFX_BUG_ON_PARANOID(rc >= n_channels);
    n_channels = rc;
-   rc = pci_enable_msix(efx->pci_dev, xentries,
-          n_channels);
   }
 
-  if (rc == 0) {
+  if (rc > 0) {
    efx->n_channels = n_channels;
    if (n_channels > extra_channels)
     n_channels -= extra_channels;
@@ -1373,11 +1376,6 @@ static int efx_probe_interrupts(struct efx_nic *efx)
    for (i = 0; i < efx->n_channels; i++)
     efx_get_channel(efx, i)->irq =
      xentries[i].vector;
-  } else {
-   /* Fall back to single channel MSI */
-   efx->interrupt_mode = EFX_INT_MODE_MSI;
-   netif_err(efx, drv, efx->net_dev,
-      "could not enable MSI-X\n");
   }
  }
 
-- 
1.7.1